### ISSUE TITLE
Internal CLR error (0x80131506) when using "struct" with DI + async operations

### ISSUE THREAD

**Issue by PawelGerr** [dotnet/runtime#48553]
### Description

There is a constellation, when using a `struct` with DI + ASP.NET Core + EF Core, which leads to an `Internal CLR error (0x80131506)`. I was able to isolate the problem and to create a repro so the issue comes almost on every run.

Repro: https://github.com/PawelGerr/ClrError

I tested the repro with Rider and with `dotnet run`. 

The solution is quite simple. If I change the `struct` to a `class` then the issue is gone.

### Further info

```
.NET SDK (reflecting any global.json):
 Version:   5.0.103
 Commit:    72dec52dbd

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.103\

Host (useful for support):
  Version: 5.0.3
  Commit:  c636bbdc8a

.NET SDKs installed:
  2.2.207 [C:\Program Files\dotnet\sdk]
  3.1.302 [C:\Program Files\dotnet\sdk]
  3.1.402 [C:\Program Files\dotnet\sdk]
  3.1.406 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]
  5.0.103 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by jkotas** [comment]
The repro crashes on checked runtime with:
```
Assert failure(PID 13892 [0x00003644], Thread: 13084 [0x331c]): Precondition failure: FAILED: IsArray() || ImplementsEquivalentInterface(ownerType.GetMethodTable()) || ownerType.GetMethodTable()->HasVariance()

CORECLR! CHECK::Trigger + 0x2C9 (0x00007ffa`657fa9e9)
CORECLR! MethodTable::GetMethodDescForInterfaceMethod + 0x473 (0x00007ffa`653240fb)
CORECLR! CEEInfo::resolveVirtualMethodHelper + 0x28C (0x00007ffa`652e2d50)
CORECLR! CEEInfo::resolveVirtualMethod + 0x16C (0x00007ffa`652e2a4c)
CLRJIT! Compiler::impDevirtualizeCall + 0xB6B (0x00007ffa`6c83e88f)
```
It may or may not be the root cause of the crash, but it should be fixed first.

**Comment by AndyAyersMS** [comment]
@jkotas if the above was from 6.0/master there is a spurious class guess enabled in the jit. Can suppress via

set COMPlus_JitGuardedDevirtualizationGuessBestClass=0

Beyond that I see the jit reporting bad IL in 

DynamicClass:ResolveService(ILEmitResolverBuilderRuntimeContext,Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope):System.Object"

Will look more deeply when I have time.

**Comment by jkotas** [comment]
Yes, I have tried it on 6.0/master. Do we have a bug on the spurious class guess leading to assert?

> If I change the struct to a class then the issue is gone.
> DynamicClass:ResolveService

This sounds like a duplicate of https://github.com/dotnet/runtime/issues/42160. @eerhardt Does it plausible that DI generates bad IL in some cases for valuetypes?

**Comment by AndyAyersMS** [comment]
> Do we have a bug on the spurious class guess leading to assert?

We do now.

**Comment by davidfowl** [comment]
We fixed an issue with value types in 5.0 I wonder if this is related.

**Comment by AndyAyersMS** [comment]
Problematic IL is
```
IL_0076  a5 13 00 00 02    unbox.any    0x2000013
IL_007b  73 14 00 00 06    newobj       0x6000014
```
Jit is looking at the newobj, which is 
```
    m_pszDebugMethodName: 0x00007ffcb9b7f398 ".ctor"
    m_pszDebugClassName: 0x00007ffcb9b6b090 "ClrIssueRepro.Dependency_1"
    m_pszDebugMethodSignature: 0x00007ffcb9b7f3c0 "instance void *(class ClrIssueRepro.TestDbContext,class ClrIssueRepro.IDependency_2,class ClrIssueRepro.IDependency_3,class ClrIssueRepro.ICulprit
```
and the arg the jit has is a struct of type `Culprit`, not a class.

**Comment by AndyAyersMS** [comment]
Above is with 6.0 -- not sure how this explains the 5.0 behavior as the bad IL should cause an invalid program exception and not an internal error. So still need to look at 5.0.

**Comment by AndyAyersMS** [comment]
5.0 checked or release build (with TC disabled) gives:
```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HM6M146Q60DL", Request id "0HM6M146Q60DL:00000005": An unhandled exception was thrown by the application.
      System.ArgumentNullException: Value cannot be null. (Parameter 'culprit')
         at ClrIssueRepro.Dependency_1..ctor(TestDbContext ctx, IDependency_2 dependency2, IDependency_3 dependency3, ICulprit culprit) in D:\bugs\r48553\ClrError\ClrError\Dependency_1.cs:line 17
         at ResolveService(ILEmitResolverBuilderRuntimeContext , ServiceProviderEngineScope )
```
but suspect under the covers this is just the .ctor call seeing bogus argument passing an happening to pick up null; some non-null value there could lead to the crashes seen with the actual product.

Looks like the jit did not detect this bad IL in 5.0. The behavior of the jit was changed in #43386.

**Comment by eerhardt** [comment]
I was able to repro the error when using the 3.1 version of DependencyInjection. Using the 5.0 and 6.0 versions of DependencyInjection I am no longer seeing an Internal CLR error. I believe it was fixed with https://github.com/dotnet/runtime/pull/42152.

I'm going to close this issue. 

Also note that using `struct`s in DI isn't fully supported yet. See https://github.com/dotnet/runtime/issues/42160. It is probably best to use `class`, especially considering you are casting the `struct` as an `interface`, which will box and allocate an object anyway. So you aren't really saving anything using a `struct`.
